my-webxr-app/src/components/PCA.tsx

```diff
@@ -0,0 +1,175 @@
+    // Lazy evaluation, checkSymmetry is false the rest of condition will not be evaluated, reducing computation.
+    if (checkSymmetry && !isSymmetricMatrix(covarianceMatrix)) {
+        throw new Error("The input matrix must be symmetric.");
+    }
+    const evd = new EVD(covarianceMatrix, { assumeSymmetric: true });
+    const S = evd.realEigenvalues;
+    S.reverse();
+    return S;
+}
+
+/**
+ * Computes the eigenvectors from a covariance matrix.
+ * 
+ * It uses Eigenvalue Decomposition (EVD) to compute the eigenvectors. The eigenvectors are then sorted in descending order 
+ * of their corresponding eigenvalues, so that the first eigenvector corresponds to the direction of maximum variance in the data, 
+ * the second to the direction of second most variance, and so on.
+ * 
+ * @param covarianceMatrix - The covariance matrix to compute the eigenvectors from.
+ * @param checkSymmetry - Optional. Whether to check if the input matrix is symmetric. Default is true.
+ * @returns A matrix of eigenvectors (U).
+ * @throws {Error} If the input matrix is not symmetric and checkSymmetry is true.
+ */
+export function computeEigenvectorsFromCovarianceMatrix(covarianceMatrix: Matrix, checkSymmetry: boolean = true): Matrix {
+    // Lazy evaluation, checkSymmetry is false the rest of condition will not be evaluated, reducing computation.
+    if (checkSymmetry && !isSymmetricMatrix(covarianceMatrix)) {
+        throw new Error("The input matrix must be symmetric.");
+    }
+    const evd = new EVD(covarianceMatrix, { assumeSymmetric: true });
+    const U = evd.eigenvectorMatrix;
+    U.flipRows();
+    return U;
+}
+
+
+/**
+ * Performs Principal Component Analysis (PCA) on the input dataset.
+ * 
+ * The function first converts the input to a Matrix if it's not already one. Then, the dataset is standardized.
+ * The covariance matrix of the standardized dataset is calculated next.
+ * The eigenvectors of the covariance matrix are computed and used to transform the dataset into the principal component space.
+ * The transformed dataset is then truncated to only include the first kComponents principal components.
+ * 
+ * @param datasetMatrix - The dataset to perform PCA on. This can be a Matrix or a data structure that can be converted to a Matrix.
+ * @param kComponents - The number of principal components to keep.
+ * @param assert - Optional. Whether to enable assertions. Default is true. Turn off to reduce computations.
+ * @returns The transformed dataset, as a first principal kComponents  Matrix.
+ * @throws {Error} If the input dataset is not a Matrix or a 2D array and assert is true.
+ */
+export function computePCA(datasetMatrix: MaybeMatrix, kComponents: number, assert: boolean = true): Matrix {
+    datasetMatrix = convertToMatrix(datasetMatrix, assert);
+    datasetMatrix = standardizeDataset(datasetMatrix, assert);
```

Comment:
We throw an error in this function, yet have no graceful error handling. Won't that just kill the process?

Reply:
oh potato, yes, i should throw a try-catch there. Thinking of disable assertion to save computations distract me.